pettingzoo/sisl/waterworld/waterworld_base.py

```diff
@@ -45,73 +50,112 @@ def velocity(self):
         assert self._sensors is not None
         return self._sensors
 
-    def sensed(self, objx_N_2, same=False):
+    def sensed(self, object_coord, object_radius, same=False):
         """Whether `obj` would be sensed by the pursuers"""
-        relpos_obj_N_2 = objx_N_2 - np.expand_dims(self.position, 0)
-        sensorvals_K_N = self.sensors.dot(relpos_obj_N_2.T)
-        sensorvals_K_N[(sensorvals_K_N < 0) | (sensorvals_K_N > self._sensor_range) | ((
-            relpos_obj_N_2**2).sum(axis=1)[None, :] - sensorvals_K_N**2 > self._radius**2)] = np.inf
+        relative_coord = object_coord - np.expand_dims(self.position, 0)
+        # Projection of object coordinate in direction of sensor
+        sensorvals = self.sensors.dot(relative_coord.T)
+        # Set sensorvals to np.inf when object should not be seen by sensor
+        distance_squared = (relative_coord**2).sum(axis=1)[None, :]
+        sensorvals[
+            (sensorvals < 0)    # Wrong direction (by more than 90 degrees in both directions)
+            | (sensorvals - object_radius > self._sensor_range)         # Outside sensor range
+            | (distance_squared - sensorvals**2 > object_radius**2)     # Sensor does not intersect object
+        ] = np.inf
         if same:
-            sensorvals_K_N[:, self._idx - 1] = np.inf
-        return sensorvals_K_N
+            # Set sensors values for sensing the current object to np.inf
+            sensorvals[:, self._idx - 1] = np.inf
+        return sensorvals
+
+    def sense_barriers(self, min_pos=0, max_pos=1):
+        sensor_vectors = self.sensors * self._sensor_range
+        sensor_endpoints = sensor_vectors + self.position
+
+        # Clip sensor lines on the environment's barriers.
+        # Note that any clipped vectors may not be at the same angle as the original sensors
+        clipped_endpoints = np.clip(sensor_endpoints, min_pos, max_pos)
+
+        # Extract just the sensor vectors after clipping
+        clipped_vectors = clipped_endpoints - self.position
+
+        # Find the ratio of the clipped sensor vector to the original sensor vector
+        # Scaling the vector by this ratio will limit the end of the vector to the barriers
+        ratios = np.divide(clipped_vectors, sensor_vectors, out=np.ones_like(clipped_vectors),
+                           where=np.abs(sensor_vectors) > 0.00000001)
+
+        # Find the minimum ratio (x or y) of clipped endpoints to original endpoints
+        minimum_ratios = np.amin(ratios, axis=1)
+
+        # Convert to 2d array of size (n_sensors, 1)
+        sensor_values = np.expand_dims(minimum_ratios, 0)
+
+        # Set values beyond sensor range to infinity
+        is_in_max_range = np.reshape((np.amax(sensor_endpoints, axis=1) < max_pos), (1, self._n_sensors))
+        is_in_min_range = np.reshape((np.amin(sensor_endpoints, axis=1) > min_pos), (1, self._n_sensors))
+        sensor_values[np.logical_and(is_in_max_range, is_in_min_range)] = np.inf
```

Comment:
I don't think I understand this, minimum ratios are always <1 because `clipped_vectors` are always less than or equal to `sensor_vectors`, so ratios < 1, aside from floating point errors. Does_sense would all be false. That being said, I don't like this code and agree it could probably be improved somehow.

Reply:
Well, the point is that the ratios are equal to 1 if they didn't get clipped and less than 1 if they did get clipped. This isn't really a necessary change, I just thought the code was a bit messy.

Reply:
Oh that makes sense, I like this change